server/src/main/java/org/apache/druid/server/coordinator/balancer/SegmentToMoveCalculator.java

```diff
@@ -0,0 +1,309 @@
+   * order to balance the segment counts of the most unbalanced datasource.
+   */
+  static int computeSegmentsToMoveToBalanceCountsPerDatasource(
+      String tier,
+      List<ServerHolder> servers
+  )
+  {
+    // Find all the datasources
+    final Set<String> datasources = servers.stream().flatMap(
+        s -> s.getProjectedSegments().getDatasourceToTotalSegmentCount().keySet().stream()
+    ).collect(Collectors.toSet());
+    if (datasources.isEmpty()) {
+      return 0;
+    }
+
+    // Compute the min and max number of segments for each datasource
+    final Object2IntMap<String> datasourceToMaxSegments = new Object2IntOpenHashMap<>();
+    final Object2IntMap<String> datasourceToMinSegments = new Object2IntOpenHashMap<>();
+    for (ServerHolder server : servers) {
+      final Object2IntMap<String> datasourceToSegmentCount
+          = server.getProjectedSegments().getDatasourceToTotalSegmentCount();
+      for (String datasource : datasources) {
+        int count = datasourceToSegmentCount.getInt(datasource);
+        datasourceToMaxSegments.mergeInt(datasource, count, Math::max);
+        datasourceToMinSegments.mergeInt(datasource, count, Math::min);
+      }
+    }
+
+    // Compute the gap between min and max for each datasource and order by largest first
+    final TreeMap<Integer, String> countDiffToDatasource = new TreeMap<>(Comparator.reverseOrder());
+    datasourceToMaxSegments.object2IntEntrySet().forEach(entry -> {
+      String datasource = entry.getKey();
+      int maxCount = entry.getIntValue();
+      int minCount = datasourceToMinSegments.getInt(datasource);
+      countDiffToDatasource.put(maxCount - minCount, datasource);
+    });
+
+    // Identify the most unbalanced datasource
+    final Map.Entry<Integer, String> maxCountDifference = countDiffToDatasource.firstEntry();
+    String mostUnbalancedDatasource = maxCountDifference.getValue();
+    int minNumSegments = Integer.MAX_VALUE;
+    int maxNumSegments = 0;
+    for (ServerHolder server : servers) {
+      int countForSkewedDatasource = server.getProjectedSegments()
+                                           .getDatasourceToTotalSegmentCount()
+                                           .getInt(mostUnbalancedDatasource);
+
+      minNumSegments = Math.min(minNumSegments, countForSkewedDatasource);
+      maxNumSegments = Math.max(maxNumSegments, countForSkewedDatasource);
+    }
+
+    final int numSegmentsToMove = maxCountDifference.getKey() / 2;
+    if (numSegmentsToMove > 0) {
+      log.info(
+          "Need to move [%,d] segments of datasource[%s] in tier[%s] to fix gap between min[%,d] and max[%,d].",
+          numSegmentsToMove, mostUnbalancedDatasource, tier, minNumSegments, maxNumSegments
+      );
+    }
+    return numSegmentsToMove;
+  }
+
+  private static int computeSegmentsToMoveToBalanceDiskUsage(
+      String tier,
+      List<ServerHolder> servers
+  )
```

Comment:
i sort of wonder if this runs into conflicts with cost balancer which primarily operates on interval adjacency instead of disk size for datasources where the segment sizes are not very homogenous. I imagine it might, but also its probably not that harmful because in the worst case it just considers more segments to move than it needed to.

Reply:
Yeah, this logic is just to determine the number of segments we would like to move, while being agnostic to the actual balancer strategy employed.
(That's not _entirely_ true since we do calculate the total time taken and thus `maxSegmentsToMove` based on time taken by `cost` strategy in a single computation. But that just means that any other existing or future strategies need to be faster than `cost`.)

Given that all historicals in a tier have the same amount of maximum disk space, we should be good. I imagine there are not many clusters out there that use heterogeneous historicals. If there are, we would have to rely on percentages but then again, that percentage would need to be somehow translated to bytes.